### ISSUE TITLE
internal/core/eval: unresolved embedded scalar disjunction influences unrelated definition

### ISSUE THREAD

**Issue by cueckoo** [cue-lang/cue#1034]
_Originally opened by @myitcv in https://github.com/cuelang/cue/issues/1034_

### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
cue version +0070bb10 linux/amd64
</pre>

### Does this issue reproduce with the latest release?

Yes

### What did you do?

```
-- foo.cue --
package foo

#SomeBaseType: {
 "a" | "b"
 #AUTO: "z"
}

V1: ("x" | "y") | *"z"
V2: ("x" | "y") | *#SomeBaseType.#AUTO
-- foo2.cue --
package foo

V3: #SomeBaseType.#AUTO
```

If we do:

```
$ cue export .
```

we get the following error:

```
V3: unresolved disjunction "a" | "b" (type string):
    ./foo2.cue:3:5
```

This seems wrong, because whilst `#SomeBaseType.#AUTO` does reference `#SomeBaseType`, the fact the disjunction is unresolved should not matter, because we are addressing `#AUTO`. 

If we then do:

```
$ cue eval foo.cue
```

(i.e. ignoring `V3`) then we get:

```
#SomeBaseType: {
    "a"
    #AUTO: "z"
} | {
    "b"
    #AUTO: "z"
}
V1: "z"
V2: "x" | "y"
```

Here we would expect `V2` to be `"z"`, but because of the error described above the default gets eliminated, hence we are left with `"x" | "y"`; so not as expected, but at least understandable in the context of the previous observation.

If we then do:

```
$ cue export foo.cue
```

we get no output, no error message, and a non-zero exit code. As stated above we would expect `V2` to be `"z"`, but given the current output from `cue eval foo.cue` we would, given the current implementation, expect an error message like:

```
V2: unresolved disjunction "a" | "b" (type string)
```

But this final scenario is effectively two errors stacked on top of each other, so it's not clear that our expectation is reasonable here.

### What did you expect to see?

As above.

### What did you see instead?

As above.

cc @dubo-dubon-duponey

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/1034#issuecomment-856686462_

CUE makes no distinction between embedded structs or embedded scalars. This behavior is especially useful if the embedded values are references.

**Comment by myitcv** [comment]
Per @mpvl this is working as expected.

**Comment by dubo-dubon-duponey** [comment]
Purely from a usage perspective this is unfortunate, as this is quite hard to find reason for.
Also, not having any output of any kind with export makes it even harder to defend...

**Comment by verdverm** [comment]
FWIW, it seems the errors are getting squashed somewhere. (also inconsistent between encoders during export)

```
$ cue export 1034.cue --out yaml
yaml: unsupported node "x"|"y" (*ast.BinaryExpr)
```

Perhaps a concreteness check could be made before exporting?

There also seems to be a bug with `eval -c` on the `foo.cue` code above

```
$ cue eval -c 1034.cue 
V1: "z"
V2: "x" | "y"
```

V2 is not concrete, so shouldn't there be an error reported?

`(using v0.4,0)`

**Comment by myitcv** [comment]
@dubo-dubon-duponey you're right, I prematurely closed this because there is at least one issue hidden in my original description. My closing the issue was more a function of the explanation regarding the referencing of `#SomeBaseType.#AUTO`. 

The issue that does need to be fixed is that `cue export foo.cue` gives no error messages but results in a non-zero exit code. This cannot be right.

@verdverm 



> Perhaps a concreteness check could be made before exporting?

Please can you raise this in a separate issue?

> V2 is not concrete, so shouldn't there be an error reported?

This seems worthy of a new issue as well if you don't mind?

**Comment by verdverm** [comment]
#1152 #1153 #1154

**Comment by mvdan** [comment]
As of ac52bda1e4b69cd1aceaaf8dbc9b51735f573ff9, the empty error message is resolved:

```
exec cue export in.cue

-- in.cue --
#SomeBaseType: {
    "a"
    #AUTO: "z"
} | {
    "b"
    #AUTO: "z"
}
V1: "z"
V2: "x" | "y"
```
```
> exec cue export in.cue
[stderr]
V2: incomplete value "x" | "y"
[exit status 1]
```

I'm fairly certain that we had that as a separately filed issue that was resolved with a test. Although I can't find it now :) In any case, it's resolved.
